euclidlib/photo/_phz.py

```diff
@@ -134,31 +140,34 @@ def _(
 
     # convert every nz into the PHZ format
     if hist:
-        # rebin the histogram as necessary
-
         # shorthand for the left and right z boundaries, respectively
         zl, zr = z[:-1], z[1:]
-
         # compute the mean redshifts
-        out["MEAN_REDSHIFT"] = np.sum((zl + zr) / 2 * nz, axis=-1) / np.sum(nz, axis=-1)
-
+        mid_z = (zl + zr) / 2
+        nz_bins = 0.5 * (nz_array[:, :-1] + nz_array[:, 1:])
+        mid_z = mid_z[-nz_bins.shape[1] :]
+        out["MEAN_REDSHIFT"] = np.sum(mid_z * nz_bins, axis=1) / np.sum(
+            nz_array, axis=1
+        )
         # compute resummed bin counts
         for j, (z1, z2) in enumerate(zip(zbinedges, zbinedges[1:])):
             frac = (np.clip(z2, zl, zr) - np.clip(z1, zl, zr)) / (zr - zl)
-            out["N_Z"][:, j] = np.dot(nz, frac)
+            frac = frac[-nz_bins.shape[1] :]
+            out["N_Z"][:, j] = np.dot(nz_bins, frac)
     else:
         # integrate the n(z) over each histogram bin
-
         # compute mean redshifts
-        out["MEAN_REDSHIFT"] = trapezoid(z * nz, z, axis=-1) / trapezoid(nz, z, axis=-1)
+        out["MEAN_REDSHIFT"] = trapezoid(z * nz_array, z, axis=-1) / trapezoid(
```

Comment:
Can you not unify these pieces code with the one above my multiply by dz if needed?